pkg/syncutil/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,94 @@
+package cachemanager
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/controller/config/process"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/logging"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/metrics"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/readiness"
+	"github.com/open-policy-agent/gatekeeper/v3/pkg/syncutil"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+var log = logf.Log.WithName("data-replication").WithValues("metaKind", "CacheManagerTracker")
+
+type CacheManager struct {
+	opa              syncutil.OpaDataClient
+	syncMetricsCache *syncutil.MetricsCache
+	tracker          *readiness.Tracker
+	processExcluder  *process.Excluder
+
+	// todo acpana -- integrate gvkaggregator
+}
+
+func NewCacheManager(opa syncutil.OpaDataClient, syncMetricsCache *syncutil.MetricsCache, tracker *readiness.Tracker, processExcluder *process.Excluder) *CacheManager {
+	return &CacheManager{
+		opa:              opa,
+		syncMetricsCache: syncMetricsCache,
+		tracker:          tracker,
+		processExcluder:  processExcluder,
+	}
+}
+
+func (c *CacheManager) AddObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	isNamespaceExcluded, err := c.processExcluder.IsNamespaceExcluded(process.Sync, instance)
+	if err != nil {
+		return fmt.Errorf("error while excluding namespaces: %w", err)
+	}
+
+	// bail because it means we should not be
+	// syncing this gvk
+	if isNamespaceExcluded {
+		// todo acpana -- consider actually calling RemoveGVKToSync in this case
+		// as we should not be tracking this GVK anymore
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	_, err = c.opa.AddData(ctx, instance)
+	if err != nil {
+		c.syncMetricsCache.AddObject(
+			syncKey,
+			syncutil.Tags{
+				Kind:   instance.GetKind(),
+				Status: metrics.ErrorStatus,
+			},
+		)
+
+		return err
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+		Kind:   instance.GetKind(),
+		Status: metrics.ActiveStatus,
+	})
+	c.syncMetricsCache.AddKind(instance.GetKind())
+
+	log.V(logging.DebugLevel).Info("[readiness] observed data", "gvk", instance.GroupVersionKind(), "namespace", instance.GetNamespace(), "name", instance.GetName())
+	return err
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	_, err := c.opa.RemoveData(ctx, instance)
+	// only delete from metrics map if the data removal was succcesful
+	if err != nil {
+		c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
```

Comment:
I sit correct that we only update the metrics if we _failed_ to remove the data?

Reply:
per offline, this is consistent with the old code (& correct i think):

https://github.com/open-policy-agent/gatekeeper/blob/b7230e0bdc3479b867acd2fd0d7192b1182c59d4/pkg/controller/sync/sync_controller.go#L203-L215

resolving but please feel free to unresolve if I misunderstood your concern here please 🙏🏼

Reply:
IIUC your version is incorrect.

It deletes the object from the metrics cache `if err != nil`.

The code snippet you attached does the opposite.  `if err != nil`, that code does a no-op (`return reconcile.Result{}`).

Reply:
🤦🏼 spot on! thanks for your patience -- reworked it; lmk if I missed the mark again!